### PR TITLE
use list of job instead of single job in chainLinker

### PR DESCRIPTION
**Pull request by marekkapowicki**

### CODE REVIEW

File: src/main/groovy/com/ofg/pipeline/core/link/AbstractPublishersFocusedJobChainLink.groovy [hunk 1]
```diff
@@ -10,27 +10,31 @@ abstract class AbstractPublishersFocusedJobChainLink<P extends Project> implemen
...
 
-    private final JobRef<P> to
+    private final List<JobRef<P>> to
     protected final boolean onSameNode
 
     protected AbstractPublishersFocusedJobChainLink(JobRef<P> to, boolean onSameNode) {
+        this([checkNotNull(to)], onSameNode)
+    }
+
+    protected AbstractPublishersFocusedJobChainLink(List<JobRef<P>> to, boolean onSameNode) {
```
Comment (cameleeck) [inline src/main/groovy/com/ofg/pipeline/core/link/AbstractPublishersFocusedJobChainLink.groovy:20]: What if you get empty list ?

**Comment by ArturGajowy** [comment]
I think we already have that capability with composedlink in a way thatvdoes not add complexity to a core abstract class. 

Also,  it makes more sense to link subsequent jobs to the first one in a forked links. If it fires a next job its more important than if its a dead-end,  so it should go first. Thus subsequent jobs should be linked to the first job in a fork, as it is in composed link.

I dont see a usecase that cant be served equally well with the dsl in its current state. Please provide one.  Untill that time I'd personally rather not see it merged.

**Comment by marekkapowicki** [comment]
I can't use existing aproaches to solve my usecase
When I use merging chains into tree the created job contains 2 sections -> "build other project"
But jenkins can use only one -> we were talking about it yesterday

So if I want to start 2 manual jobs i can
- make a dirty hack and run one auto (which triggers manual)  and one manual job. This solution does not change the jenkins-pipeline-dsl
- change the jenkins-pipeline-dsl and run many jobs (comma separated) in "Downstream Project Names"

This PR allows to realize second solution

**Comment by ArturGajowy** [comment]
I think the comma separated jobs wont work because arrows for trigerring builds won't be displayed b/c of limitations of the dashboard (or some other) plugin. I mentioned that in our talk :) Make sure you check that for a new build, not one that had any of the manual jobs run already (unless that changed,  somehow). I think @szpak or @tulinski will be able to confirm or debunk what I'm saying.

**Comment by ArturGajowy** [comment]
Hi! Any progress on that? I'm curious if and how the issue was solved :)

**Comment by s2lomon** [comment]
So it seems that it's a bit more complicated. Everything works well, as long as we stick to many auto and one manual steps. Unfortunately, when you try to use many manual steps, the only supported configuration is with one manual trigger, able to trigger many jobs manually. 
Due to that, I think that we have 3 options:
1. repair the manual triggering so that it could work as auto 
2. hack it around and try to merge all manual triggers into single trigger on Node level...
3. make our dsl explicit about what's possible and what's not.

**Comment by ArturGajowy** [comment]
@ 1: good luck with fixing the involved jenkins plugins ;)
@ 2: possible, but a bit complex. Additional postprocessing step would be needed either on XML or joblinks level (merging manual links from same job before using them for job configuraiton)
@ 3: would rather not. It's a bug in jenkins and its plugins, that thing should just work. Once they fix the bug, the will need to change again.

Do I get that right that you guys are still reluctant to apply the prepend-manual-link-with-a-dummy-auto-job workaround? Aren't there more interesting problems to solve than how to build support for hiding a jenkins bug from our not-so-numerous users?

I've heard from @marekkapowicki that he's got a fix in delivery-pipeline plugin for those manual triggers. Ask @szpak, but I think that bug had already been reported at least half a year ago and their response was something like "it's not a bug, it's a feature" or "your request is against the pipeline semantics". 

Also, you're already depending on my fork of delivery-pipeline that has a change they're not willing to merge 'because semantics'. I'm fine with you further diverging from the mainline, but are **you**? ;)

**Comment by s2lomon** [comment]
@ArturGajowy regarding this bug, it's probably not a but. The thing is that xmls that are defining jobs does not have any xsds that would limit their structure (obviously) , However when you would try to add a manual or auto trigger from UI (so as it's been always done) you would find yourself in a place where you can add only one auto and only one manual trigger, with multiple downstream projects separated with a comma.

I would rather say that this is a coincidence that auto triggers are working right now and manual step behavior is more consistent... this might hit us on the go. 

Regarding our free time - we have RnD, so chill out ;)

**Comment by ArturGajowy** [comment]
@S2lomon Well, when stated that way it makes sense to limit the DSL so that at each step only one trigger of a given type is possible. This will be especially fun for Downstream Paremeterized trigger, which is actually a multi-trigger (and each triggered downstream job can have different parameters config).

Or one could argue that jenkins is messed up at its basis then... ;) I find this gui-only-limitation (only one downstream trigger of given type per job) quite arbitrary.

Great to hear you've got a RnD day! ;) Use it for whatever rocks your boat ;> I'm a bit ashamed of how much time we've spend discussing this, given the simplicity of the workaround and the significance of this for our lives... :D

**Pull request closed without merging**